custom-schemes.md

```diff
@@ -2,40 +2,54 @@
 
 Existing specifications like [OpenID4VP](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#name-wallet-invocation) and [ISO 18013-7](https://www.iso.org/standard/82772.html) currently define the use of custom URL schemes (eg. `openid4vp:` and `mdoc:`) to enable verifier websites to invoke holder (wallet) applications. This approach works in some scenarios, but has a number of problems we aim to solve by designing a browser API which can replace the use of custom URL schemes in these specifications. This DRAFT document aims to outline the questions and concerns with the use of custom URL schemes for identity presentment in order to provide more detail on potential benefits of a dedicated browser API.
 
-1. Can wallets reliably determine their invoker?
+## Can wallets reliably determine their invoker?
 
-   - Android and iOS offer no facility to determine the web origin which triggered a navigation to a custom URL scheme. 
+   - Android and iOS offer no facility to determine the web origin which triggered a navigation to a custom URL scheme.
    - This seems to increase the opportunity for attacks such as replay attacks, man-in-the-middle attacks and phishing. It also seems to limit the opportunity for abuse detection and prevention.
    - A browser API can securely pass the requesting origin on to the wallet so that the wallet can know with confidence what origin the user was on when the request was initiated.
+   
+## How are wallet apps supposed to pass info back to the invoking page?
 
-2. Can wallets limit requests to secure contexts?
+Does the page give the wallet a URL to navigate back to as part of the custom scheme URL it uses to launch the wallet?
 
-   - If a user uses a website via http (not https) and it invokes a custom scheme, it's possible that an attacker with access to the network could eavesdrop on or even modify the request (such as by replacing the verifier's public key with their own). 
-   - A browser API could take the position that it will function only within the context of a secure context (https).
+If so, that's a problem because it won't go back to the same tab. Apps don't have the ability to return information to the same page that invoked them after a custom scheme invocation; they can just open a new page in a new tab.
```

Comment:
Well, for some definition of "definitely work". :) It is possible to end in a "successful state", but that ignores the UX challenges of what happens when things go wrong. In the best case (on a same-device flow), presuming that the protocol has some sort of "navigate to this URL when you're done" feature, you're left with a "dead tab" in the browser -- something that the individual will never go back to and have to clean up later.

One of the design goals of [CHAPI](https://chapi.io/) was to not "break the flow" in the existing web page (and not to leave "dead tab"s around). This is also aligned w/ the design goals for Payment Request (and Payment Handler). That is, you want the browser API to always tell you when the interaction is done (success or failure) and return you to the exact place you were on the website that invoked the interaction in the first place. I expect FedCM has the same (ideal world) requirements.

There are at least three problems w/ invoking a native app or a web-app (using what's available today) in a separate window:

1. In a same-device flow, when the flow completes successfully, navigating back to the same tab/window that the individual was in (in the browser) is not possible. In the best case, you have to open a new tab/window, leaving "dead tab"s around. 
2. In a same device-flow, the UX is not ideal -- popping windows up (because you need to be in first party mode in many cases, and sometimes you need to log in -- which pops up more windows if you are using an external IdP) can cause the user to feel as if something bad is happening - "User: WTF is going on here? This doesn't feel right *closes window, breaking the interaction*").
3. In a multi-device or same-device flow, the user is never navigated back to the originating page when there is an error or the interaction is canceled. They're left in the "invoked app" and have to manually switch back to their browser session, which gets no indication that things went wrong and therefore "hangs" for an unacceptably long time -- 5 minutes, 10 minutes... how long should a site give an individual to complete an out-of-band flow?

It's not just a developer ergonomics issue, it's a UX issue, both for same-device and multi-device flows.

Reply:
> In a same-device flow, when the flow completes successfully, navigating back to the same tab/window that the individual was in (in the browser) is not possible. In the best case, you have to open a new tab/window, leaving "dead tab"s around.

Definitely agree not having a way to navigate back to the browser app is a problem that we should add to this list.

I had assumed the "best case" wasn't that the wallet would open a new browser window (which browser?), but that it would send a message to a server which the existing web page was waiting on and the user would manually switch back to their last application and see it complete. But there are problems with that as well (eg. under memory pressure the browser may have been killed).

The jarring UX of a full screen app switch is also a problem I can add to the list. I think I've already captured your "3".

Reply:
> the user would manually switch back to their last application and see it complete

This is a tall UX ask of the user. Some of the testing we performed showed some individuals not understanding what the next step in the process was... so, after sending the presentation, they just sat there looking at the wallet, not knowing what to do (because some of them thought they were still in the browser and others didn't know about the usage pattern where you complete something in an app and then switch back to a browser -- it's a rare usage pattern on mobile phones to bounce from browser->app->browser for the same logical flow).

Some were expecting wallet to tell them the next step, and in some cases, the wallet just said "transaction complete!" and then nothing. This could be mitigated by the wallet showing a UI telling them to switch back to their browser (but that feels like a UX failure). To get a consistent experience across all wallets for that UX flow, you'd have to get all wallets to implement that (hacky) UX hint. We've also found that in some cases, once the individual is in the wallet, the wallet software wants them to do other things and keep them there ("Get another coupon!" ... "Add a credit card!" ... "Wait, update your password!" ... "Turn on MFA!")

So yes:

1. Full-screen is jarring, a modal that allows the UX flow to be performed and then return the individual back to the page (with a concrete transaction result for the website to process) is better.
2. Giving the wallet some hint about where to navigate the user back to (if we don't do the modal thing above) might also be helpful.

We might consider something like the "peek" feature in Arc as the ideal UX pattern to follow here: 

https://www.reddit.com/r/ArcBrowser/comments/113ui8u/peek_by_arc_the_best_tab_never_made/